### PR TITLE
Go 1.21.6, 1.21.7, 1.21.8, 1.22.0, 1.22.1

### PR DESCRIPTION
**Pull request by Fabio1988**
I tried to follow last update, hopefully everything covered

also removed go 1.20 version to not exceed 360 minutes building time of github actions

this needs https://github.com/crazy-max/goxx/pull/70

### CODE REVIEW

**Review by mochaaP — APPROVED** [review]
LGTM

**Comment by mochaaP** [comment]
cc @crazy-max

**Comment by mochaaP** [comment]
I believe those are false positives? Rerun should pass the test

**Comment by crazy-max** [comment]
> I believe those are false positives? Rerun should pass the test

I don't think so: https://github.com/crazy-max/xgo/actions/runs/8278490599/job/22671620226?pr=123#step:5:200

```
0.180 INFO: Starting xgo/69d0307
0.180 DBG: config: &{Repository:. Package: Prefix:test Remote: Branch: Dependencies: Arguments: Targets:[*/*]}
0.180 DBG: flags: &{Verbose:false Steps:false Race:false Tags: LdFlags: Mode:default VCS:true TrimPath:false}
0.180 INFO: Don't use go modules (go.mod not found)
0.180 INFO: Cross compiling . package...
0.185 Building locally ....
0.352 Compiling for linux/amd64...
0.362 go: modules disabled by GO111MODULE=off; see 'go help modules'
0.363 ERROR: Failed to cross compile package: exit status 1.
```

Might need changes in xgo since Go 1.22

**Comment by mochaaP** [comment]
<https://go.dev/doc/go1.22#go-command>

tl;dr: `go get` is no longer supported outside of a module in the legacy `GOPATH` mode (that is, with `GO111MODULE=off`)

**Comment by Fabio1988** [comment]
might be related to this commit? https://github.com/crazy-max/xgo/commit/607610c013ca0ee9d65e8e9bb87fa2ba76fb2f56

**Comment by mochaaP** [comment]
We might need to refactor the `go get` flow here.

**Comment by Fabio1988** [comment]
Sorry I have no clue to achieve this, I thought it might be easy to update version to get my application workflow executable again. I updated already to 1.22. I guess I need to wait then :)

**Comment by mochaaP** [comment]
🤔 I'll take a look later.

**Comment by mochaaP** [comment]
Note: need to somehow add ghq to the base image. (either through Alpine's testing repo or `go install`)

```patch
diff --git a/rootfs/usr/local/bin/xgo-build b/rootfs/usr/local/bin/xgo-build
index 1780139..50de89f 100755
--- a/rootfs/usr/local/bin/xgo-build
+++ b/rootfs/usr/local/bin/xgo-build
@@ -99,7 +99,7 @@ else
 
   # Otherwise download the canonical import path (may fail, don't allow failures beyond)
   echo "Fetching main repository $1..."
-  GO111MODULE=off go get -v -d "$1"
+  GHQ_ROOT=$GOPATH_ROOT ghq get "$1"
   set -e
 
   cd "$GOPATH_ROOT/$1"
diff --git a/tests/c/go.mod b/tests/c/go.mod
new file mode 100644
index 0000000..8740bd4
--- /dev/null
+++ b/tests/c/go.mod
@@ -0,0 +1,3 @@
+module tests/c
+
+go 1.22.1
diff --git a/tests/cpp/go.mod b/tests/cpp/go.mod
new file mode 100644
index 0000000..ba3cd35
--- /dev/null
+++ b/tests/cpp/go.mod
@@ -0,0 +1,3 @@
+module tests/cpp
+
+go 1.22.1
```

**Comment by Fabio1988** [comment]
> Note: need to somehow add ghq to the base image. (either through Alpine's testing repo or `go install`)

so we need to adapt goxx image?

**Comment by Fabio1988** [comment]
do you guys have a discord server for discussion or contact possibilities? also willed to donate to you :) 
not sure about the effort to do, but I would like to use golang 1.22 in production :smile:

**Comment by mochaaP** [comment]
Will try to find a time to implement this. I think it's fine to `go install` it at a separate layer from the first build stage?

**Comment by Fabio1988** [comment]
> Will try to find a time to implement this. I think it's fine to `go install` it at a separate layer from the first build stage?

But do I think the right way that this needs fixing in goxx? Or what do you mean with go install? `go install github.com/x-motemen/ghq@latest`?

**Review by mochaaP — CHANGES_REQUESTED** [review]
```diff
diff --git a/Dockerfile b/Dockerfile
index 0a72b9a..238e25e 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -3,7 +3,7 @@
 ARG GO_VERSION="1.22.1"
 ARG OSXCROSS_VERSION="11.3"
 ARG XX_VERSION="1.3.0"
-ARG ALPINE_VERSION="3.18"
+ARG ALPINE_VERSION="3.19"
 ARG PLATFORMS="linux/386 linux/amd64 linux/arm64 linux/arm/v5 linux/arm/v6 linux/arm/v7 linux/mips linux/mipsle linux/mips64 linux/mips64le linux/ppc64le linux/riscv64 linux/s390x windows/386 windows/amd64"
 
 FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
@@ -13,6 +13,11 @@ ENV CGO_ENABLED=0
 RUN apk add --no-cache file git
 WORKDIR /src
 
+FROM base AS ghq
+ARG GHQ_VERSION="1.5.0"
+RUN --mount=type=cache,target=/go/pkg/mod \
+  go install github.com/x-motemen/ghq@v${GHQ_VERSION}
+
 FROM base AS version
 RUN --mount=target=. \
   echo $(git describe --match 'v[0-9]*' --dirty='.m' --always --tags) | tee /tmp/.version
@@ -100,6 +105,7 @@ EOT
 FROM crazymax/osxcross:${OSXCROSS_VERSION} AS osxcross
 FROM goxx-base
 COPY --from=build /usr/bin/xgo /usr/local/bin/xgo
+COPY --from=ghq /go/bin/ghq /usr/local/bin/ghq
 COPY --from=osxcross /osxcross /osxcross
 
 ENV XGO_IN_XGO="1"
diff --git a/rootfs/usr/local/bin/xgo-build b/rootfs/usr/local/bin/xgo-build
index 1780139..50de89f 100755
--- a/rootfs/usr/local/bin/xgo-build
+++ b/rootfs/usr/local/bin/xgo-build
@@ -99,7 +99,7 @@ else
 
   # Otherwise download the canonical import path (may fail, don't allow failures beyond)
   echo "Fetching main repository $1..."
-  GO111MODULE=off go get -v -d "$1"
+  GHQ_ROOT=$GOPATH_ROOT ghq get "$1"
   set -e
 
   cd "$GOPATH_ROOT/$1"
diff --git a/tests/c/go.mod b/tests/c/go.mod
new file mode 100644
index 0000000..8740bd4
--- /dev/null
+++ b/tests/c/go.mod
@@ -0,0 +1,3 @@
+module tests/c
+
+go 1.22.1
diff --git a/tests/cpp/go.mod b/tests/cpp/go.mod
new file mode 100644
index 0000000..ba3cd35
--- /dev/null
+++ b/tests/cpp/go.mod
@@ -0,0 +1,3 @@
+module tests/cpp
+
+go 1.22.1
```

**Comment by mochaaP** [comment]
Test runs: <https://github.com/mcha-forks/xgo/actions/runs/8865942312/job/24342766269>

**Comment by Fabio1988** [comment]
> Test runs: https://github.com/mcha-forks/xgo/actions/runs/8865942312/job/24342766269

Thanks a lot.. applied diff and cleaned up commits :)

**Comment by mochaaP** [comment]
cc'ing @crazy-max for this.

**Comment by mochaaP** [comment]
@Fabio1988 
> also willed to donate to you :)

If you found my works useful, please consider donating to sekai instead: <https://github.com/sponsors/nekohasekai>. `sing-box` help me stay connected to the uncensored internet and made my contributions possible. :D

**Comment by Fabio1988** [comment]
> @Fabio1988
> 
> > also willed to donate to you :)
> 
> If you found my works useful, please consider donating to sekai instead: https://github.com/sponsors/nekohasekai. `sing-box` help me stay connected to the uncensored internet and made my contributions possible. :D

Sure done, already sponsored crazy-max as well. Great work :)

**Comment by Fabio1988** [comment]
any chance we can run the github actions on this change pls? should be safe regarding to https://github.com/mcha-forks/xgo/actions/runs/8865942312/job/24342766269 :) @crazy-max

File: tests/c/go.mod [hunk 1]
```diff
@@ -0,0 +1,3 @@
+module tests/c
+
+go 1.22.1
```
Comment (crazy-max) [inline tests/c/go.mod:3]: Don't think patch is required, can we keep major/minor?

```suggestion
go 1.22
```

File: tests/cpp/go.mod [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module tests/cpp
+
+go 1.22.1
```
Comment (crazy-max) [inline tests/cpp/go.mod:3]: Same here

```suggestion
go 1.22
```

File: Dockerfile [hunk 3]
```diff
@@ -1,18 +1,23 @@
...
 FROM --platform=$BUILDPLATFORM tonistiigi/xx:${XX_VERSION} AS xx
-FROM --platform=$BUILDPLATFORM golang:1.20-alpine${ALPINE_VERSION} AS base
+FROM --platform=$BUILDPLATFORM golang:1.22-alpine${ALPINE_VERSION} AS base
 COPY --from=xx / /
 ENV CGO_ENABLED=0
 RUN apk add --no-cache file git
 WORKDIR /src
 
+FROM base AS ghq
+ARG GHQ_VERSION="1.5.0"
```
Comment (crazy-max) [inline Dockerfile:17]: Could you move this ARG in the global space? (right after OSXCROSS_VERSION).

Also can be updated to latest stable `1.6.1`: https://github.com/x-motemen/ghq/releases/tag/v1.6.1

Comment (crazy-max) [inline Dockerfile:17]: Could you move this ARG in the global space? (right after OSXCROSS_VERSION). Make things easier to update.

Also can be updated to latest stable `1.6.1`: https://github.com/x-motemen/ghq/releases/tag/v1.6.1

```suggestion
ARG GHQ_VERSION
```

File: rootfs/usr/local/bin/xgo-build [hunk 4]
```diff
@@ -99,7 +99,7 @@ else
 
   # Otherwise download the canonical import path (may fail, don't allow failures beyond)
   echo "Fetching main repository $1..."
-  GO111MODULE=off go get -v -d "$1"
+  GHQ_ROOT=$GOPATH_ROOT ghq get "$1"
```
Comment (crazy-max) [inline rootfs/usr/local/bin/xgo-build:102]: We need to ensure `ghq` is available in PATH before using it.

Comment (Fabio1988) [reply rootfs/usr/local/bin/xgo-build:102]: Do I need to change something to achieve it? As ghq is installed in Dockerfile - shouldn't it be available in PATH at this point?

**Comment by Fabio1988** [comment]
Thanks for the review, will Squash commits later

File: Dockerfile [hunk 5]
```diff
@@ -1,18 +1,24 @@
...
+FROM --platform=$BUILDPLATFORM golang:1.22-alpine${ALPINE_VERSION} AS base
 COPY --from=xx / /
 ENV CGO_ENABLED=0
 RUN apk add --no-cache file git
 WORKDIR /src
 
+FROM base AS ghq
+ARG GHQ_VERSION
+RUN --mount=type=cache,target=/go/pkg/mod \
+  go install github.com/x-motemen/ghq@v${GHQ_VERSION}
```
Comment (Fabio1988) [inline Dockerfile:20]: @crazy-max isn't ghq available after this command (my answer is related to your other question) :)

File: Dockerfile [hunk 6]
```diff
@@ -100,6 +106,7 @@ EOT
 FROM crazymax/osxcross:${OSXCROSS_VERSION} AS osxcross
 FROM goxx-base
 COPY --from=build /usr/bin/xgo /usr/local/bin/xgo
+COPY --from=ghq /go/bin/ghq /usr/local/bin/ghq
```
Comment (Fabio1988) [inline Dockerfile:109]: and then finally with this ? Otherwise the Docker build would fail, no?

File: rootfs/usr/local/bin/xgo-build (hunk 4 above)
Comment (crazy-max) [reply rootfs/usr/local/bin/xgo-build:102]: This is just a nit to make sure it's available. I'm fine if we left this check.

**Comment by Fabio1988** [comment]
thanks so the last piece was that the test files used go 1.22 instead of 1.21? does this even matter for the tests? :)

**Comment by crazy-max** [comment]
> thanks so the last piece was that the test files used go 1.22 instead of 1.21? does this even matter for the tests? :)

Yes it matters as 1.22 toolchain is not available with Go 1.21.
